### ISSUE TITLE
Reconsider use of global Vuex

### ISSUE THREAD

**Issue by subdavis** [girder/girder_web_components#83]
After speaking with @matthewma7 about the future of GWC and plans for the Girder 3 UI, I wanted to revisit the decision to not use a global Vuex instance in girder web components.

This issue assumes GWC will be used to build Girder 3 UI.  If that isn't true, there are still cases where a global vuex store is useful.

Here's an example of how a global store could be unobtrusive and easy to integrate into third party apps.

```javascript
// src/store/index.js

export const module = {
  state: { ... },
  mutations: { ... },
  actions: { ... },
  getters: { ... }
}

export default new Vuex.Store({
  modules: { default: module },
});
```

... then the app could either use our store or construct its own with a distinct namespace for GWC.  If needed, the app could have multiple GWC namespaces so that GWC components could be isolated from each other (this should be rare).  Each GWC component should take an optional namespace prop for indicate which store to bind to.

Consumer apps could also decorate or extend the vuex store, removing the need for "preUpsert" and "postUpsert" hooks which I was hesitant to add in the first place.

Some problems I see with the current approach

* sharing and mutating state can be messy.  Any part of an application could want access to the current browser location or directory contents.  Existing application structure may not be optimized for this need, so you could end up with `location.sync` through many levels of components.
* sharing data between GWC components isn't currently possible.   If modules A and B both want file or folder details, for example, they have to make separate requests to girder.  This is fine when girder is co-located.  I've been testing with network throttling enabled recently and this issue becomes more noticeable.   A more concrete example is building the breadcrumb in both the browser and new-folder widget.
* state mutations (such as "rename a folder") that originate from outside GWC are hard to capture.

I can see these things negatively impacting code quality and UX in the future, both for GWC and third-party consumers.

I'm fairly convinced that we should emulate the structure over at https://github.com/gothinkster/vue-realworld-example-app .  I know this gets tricky because we can't completely hide the rest service from the components.  There will inevitably be uses that aren't captured through a vuex commit/action api, but it should be treated as more of a fallback.

Rather than `injext: [ "girderRest" ]` components should have to go through the vuex store for (most) mutations in order to provide for cleaner reactivity through the app.  That `refresh()` in DataBrowser has been weighing heavy on me since I put it in -- it's a bad pattern and one that we would have to replicate in every stateful component. 

TL;DR I believe that all of my code sins so far (`refresh()` and `postUpsert()`, among others) are due to trying to avoid vuex, and without it, there will surely be more sins to follow as we build more complex apps and state becomes more unwieldy.

**Comment by matthewma7** [comment]
Just to clarify, I and @subdavis talked about GWC and I shared following opinions for using GWC in custom apps or potential Girder 3 client.
- Monkey patching Vue component like what we do with wrapping backbone's render is still feasible and practically needed. I have done that in some projects. It allows writing upstream library/code without worrying about all possible quirky needs of downstream code.
- I am fond of the statelessness (no Vuex) current GWC components, it's straight forward to use.
- Introduce presentational and container components, but it seems this idea is not so relevant to the point @subdavis talked about here.

IMHO, if we ask a downstream app to **explicitly** use Vuex, it is a lot to ask (from the downstream code and developer) to learn outside of Standard Vue.js. Instead, I think it would be better to setup something **implicit** during the Vue.install(Girder), to set up a communication bus link or state storage for all GWC component. So a downstream app will use GWC component as normal Vue.js component, but the GWC component could talk to each other behind the scenes to do caching in your bullet point 2 and even propagation in bullet point 3. For URL management, I didn't think it through yet, at this moment.

PS: I kinda think caching is an overkill for human triggered stuff like the breadcrumb, but that's a totally separate topic)

**Comment by subdavis** [comment]
> if we ask a downstream app to **explicitly** use Vuex, it is a lot to ask

I disagree. There's nothing mandatory for them to learn.  We are asking downstream to add 2 boilerplate lines that any Vue developer should be familiar with.   We already make them add 8ish to initialize girder and girderRest.

```javascript
import store from '@girder/compnents/src/store';

// ... other init

new Vue({
    store,  // here's their new line
    render: h => h(App),
    provide: { girderRest },
  }).$mount('#app');
```

> Instead, I think it would be better to setup something **implicit** during the Vue.install(Girder), to set up a communication bus link or state storage for all GWC component.

I think that's a big sacrifice to make to avoid those 2 extra lines.  I don't think we should make up our own new pattern, especially since upstream developers also derive value from having optional access to vuex.

> I am fond of the statelessness (no Vuex) current GWC components,

Without vuex, we put the burden of syncing state on the upstream developer.  Vuex would actually allow downstream developers to **care less** about the internals of GWC.  If they include componentA and componentB, they just work, rather than having to broker events between the two.

> Caching

I'm not trying to make a case for caching, I'm trying to avoid multiple components all firing the exact same request because each is unaware of the other.

Currently, if a state change in one component should cause a refresh in another, the downstream developer has to propagate that event through their own app, then everybody has to fire a bunch of duplicate requests to get the same data. Adding vuex would actually make the GWC components seem _more_ self contained.

Current behavior:
component causes mutation (like a folder name change) -> fires events to refresh the view -> emit event to upstream container -> upstream has to pass that event up or down the tree to other GWC -> other GWC fire the same requests to update their own views _and_ the upstream app has to fire more duplicate requests because it can't read GWC's state.

With vuex:
component causes mutation -> single round of requests in action handler -> vuex update causes re-render everywhere that cares about this, and upstream app can also react to this change.

**Comment by zachmullen** [comment]
Might be my lack of experience showing here, but how do we expect that an out-of-the-box store will be able to coordinate state between *arbitrary* compositions of GWC components? Could we provide an example with some store state & actions?

**Comment by matthewma7** [comment]
> Rather than `injext: [ "girderRest" ]` components should have to go through the vuex store for (most) mutations in order to provide for cleaner reactivity through the app.

I thought you meant more than just two line change becuase you said
Could you give an example of that?


> component causes mutation (like a folder name change) -> fires events to refresh the view -> emit event to upstream container -> upstream has to pass that event up or down the tree to other GWC -> other GWC fire the same requests to update their own views _and_ the upstream app has to fire more duplicate requests because it can't read GWC's state.

I didn't suggest we keep the current behavior, I said we do something _implcit_ and _smart_ inside the Vue.install(Girder), where we setup what ever mechanism for those event progagation.

**Comment by matthewma7** [comment]
If what I am saying about implicit and smart is not clear, I can create a proof of concept around the `DataTable.refresh()` 🙂

**Comment by subdavis** [comment]
@matthewma7 are you proposing some kind of private event bus between GWC components that would basically broadcast refresh events?  I don't know if that's the final solution, but it would be a lot better than what we have right now.  If I'm misunderstanding, maybe you could provide an example?  Thanks!

@zachmullen that's exactly the right question and not one I've reached a complete solution on.  To borrow a phrase from Chris K, I expect the 80% case to be that GWC is composed in the same way every time.

Here's an example I think will solve 80%:

```vue
<template>
<v-app>
    <girder-search>  // when search result is clicked, change browser location to parent folder of selected item.
    <girder-data-browser/>  // with no location specified, commit location updates to the vuex state
    <girder-description/>  // by default, show description of the currently selected row item
</v-app>
</template>
```

All of that happens behind the scenes and the upstream doesn't have to care about any of it.  Location, Selected, Pagination, and Rows all live in vuex.

If they want to listen to state changes or use the current location in another part of the app, they can use `mapState` (@matthewma7 you're right that it can involve more than just those 2 lines, but it doesn't have to).  This is all about providing reasonable defaults.

Every app I've created that uses GWC so far (I've made 4 now) belongs to this use case.  All I've wanted to know is when a user clicks and selects items in the data browser.  

The 20% case could also be pretty simple, and doesn't necessarily require vuex knowledge. 

```vue
<template>
<v-app>
    <girder-data-browser :location.sync="location"/>  // specify location, because upstream wants to control it.  Girder Data Browser will ignore the vuex browser location.
    <girder-description :item="directoryVideo"/>
</v-app>
</template>

<script>
export default {
  data() {
    return { location: null };
  },
  computed: {
    directoryVideo() {
       // given the current location, search for and return an item that looks like a video.
       return getVideoFromLocation(this.location);
    },
  },
}
</script>
```

In this case, vuex still contains Rows, Selected, Pagination.  These are accessible through vuex to the whole app.

Finally, that example above could also be implemented with explicit knowledge of the vuex store:

```vue
<template>
<v-app>
    <girder-data-browser/>  // location lives in vuex again.
    <girder-description :item="directoryVideo"/>current location
</v-app>
</template>

<script>
import { mapState } from 'vuex';
export default {
  data() {
    return { location: null };
  },
  computed: {
    ...mapState(["browserLocation"]),
    directoryVideo() {
       // given the current location, search for and return an item that looks like a video.
       return getVideoFromLocation(this.browserLocation);
    },
  },
}
</script>
```

**Comment by matthewma7** [comment]
@subdavis Thank you for the example. now I see the picture you are envisioning --- GWC has it's a whole state tree alongside the downstream app. I have a question, I think Vuex has a **single** source of truth design, so would it require a context concept if there are multiple <girder-data-browser/> or <girder-description> in the app? 

> @matthewma7 are you proposing some kind of private event bus between GWC components that would basically broadcast refresh events? I don't know if that's the final solution, but it would be a lot better than what we have right now. If I'm misunderstanding, maybe you could provide an example? Thanks!

Yes, you understand me correctly. but I would create a simple proof of concept around the DataTable.refresh().

**Comment by zachmullen** [comment]
> I think Vuex has a **single** source of truth design

This was my understanding as well, but I'm curious: are there other component libraries that use vuex in the way you're proposing @subdavis ? Would be nice to follow someone else's lead, especially if there's a well-established pattern.

**Comment by subdavis** [comment]
Yep!  If there are multiple sets, the upstream would have to initialize multiple namespaces of the same GWC store and pass a context to the component.  I would call the default "gwc" and take it as an optional prop.

I'm willing to bet that this will be extremely rare, and so a somewhat degraded developer experience is worth it (to me) for a much better one in the average case.

**Comment by matthewma7** [comment]
@subdavis I created a proof of concept of the inter component event I was talking about in place of DataBrowser.refresh() https://github.com/girder/girder_web_components/pull/90

**Comment by jbeezley** [comment]
One of the original reasons for not going with Vuex modules was a lack of prior art in the Vue ecosystem.  It looks like Vuetify is going to start using Vuex modules in it's components.  The only example I could find so far is <https://cognito.vuetifyjs.com/>.  It is very simple, but it does look a lot like the girder axios session object we are creating in GWC, but in Vuex form.  We should keep an eye on how this transition goes for Vuetify.
